### PR TITLE
Rework

### PR DESCRIPTION
**Pull request by Misfiy**

### CODE REVIEW

File: AdminTools/Commands/HintBroadcast/User.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using CommandSystem;
+
+namespace AdminTools.Commands.HintBroadcast
+{
+    internal class User
+    {
+    }
+}
```
Comment (VALERA771) [inline AdminTools/Commands/HintBroadcast/User.cs:14]: Why we even need it then?

**Review by VALERA771 — COMMENTED** [review]
1. Please don't make commants with first capital letter
2. Everything else seems good, but putting `IUsageProvider` in childrent command is kinda useless

File: AdminTools/Commands/HintBroadcast/User.cs (hunk 1 above)
Comment (Misfiy) [reply AdminTools/Commands/HintBroadcast/User.cs:14]: Yamato was gonna seperate them
And that's the plan for the future, but currently doing other stuff

**Comment by Misfiy** [comment]
> 1. Please don't make commants with first capital letter
What do you mean commants?

**Comment by VALERA771** [comment]
> > 1. Please don't make commants with first capital letter
> 
> What do you mean commants?

Check AdminTools/Commands/InstantKill/Add.cs for example. There is also some other commands with such things

> > 2.Everything else seems good, but putting IUsageProvider in childrent command is kinda useless
> 
> Does it not work?

Only if you want to use DisplayCommandUsage extensions. Command just won't appear like a normal command

**Review by VALERA771 — APPROVED** [review]
Merge time 😊

File: AdminTools/Commands/Grenade.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+            foreach (Player player in players)
+            {
+                if (player.IsDead)
+                    continue;
+                if (Projectile.CreateAndSpawn(type, player.Position, player.Rotation).Is(out TimeGrenadeProjectile timeGrenadeProjectile) && fusetime.HasValue)
+                    fusetime = timeGrenadeProjectile.FuseTime = fusetime.Value;
+            }
+
+            response = $"{type} has been sent to the followed player{(fusetime.HasValue ? $". Will explode in {fusetime}s" : string.Empty)}: \n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Grenade.cs:62]:
```suggestion
            response = $"A grenade ({type}) has been sent to the following player: {(fusetime.HasValue ? $". The grenade will explode in {fusetime} seconds." : string.Empty)}: \n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/InstantKill/InstantKill.cs [hunk 3]
```diff
@@ -1,145 +1,42 @@
...
 
 namespace AdminTools.Commands.InstantKill
 {
     [CommandHandler(typeof(RemoteAdminCommandHandler))]
     [CommandHandler(typeof(GameConsoleCommandHandler))]
-    public class InstantKill : ParentCommand
+    public class InstantKill : ParentCommand, IUsageProvider
     {
-        public InstantKill() => LoadGeneratedCommands();
+        public override string Command { get; } = "instantKill";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/InstantKill/InstantKill.cs:11]:
```suggestion
        public override string Command { get; } = "instantkill";
```

File: AdminTools/Commands/InstantKill/List.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace AdminTools.Commands.InstantKill
+{
+    public class List : ICommand
+    {
+        public string Command { get; } = "list";
+
+        public string[] Aliases { get; } = Array.Empty<string>();
+
+        public string Description { get; } = "Get instantkill list player";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/InstantKill/List.cs:16]:
```suggestion
        public string Description { get; } = "Get a list of every player with InstantKill enabled.";
```

File: AdminTools/Commands/InstantKill/List.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            if (!sender.CheckPermission("at.instakill"))
+            {
+                response = "You do not have permission to use this command";
+                return false;
+            }
+
+            StringBuilder playerLister = StringBuilderPool.Pool.Get();
+
+            playerLister.Append(Main.InstantKill.Any() ? "Players with instant killing on:\n" : "No players currently online have instant killing on");
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/InstantKill/List.cs:28]:
```suggestion
            playerLister.Append(Main.InstantKill.Any() ? "Players with InstantKill enabled:\n" : "No players currently online have instant killing on");
```

File: AdminTools/Commands/InstantKill/Remove.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace AdminTools.Commands.InstantKill
+{
+    internal class Remove : ICommand, IUsageProvider
+    {
+        public string Command { get; } = "remove";
+
+        public string[] Aliases { get; } = Array.Empty<string>();
+
+        public string Description { get; } = "remove instantkill to this player";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/InstantKill/Remove.cs:16]:
```suggestion
        public string Description { get; } = "Disable InstantKill from this player.";
```

File: AdminTools/Commands/InstantKill/Remove.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            }
+
+            for (int i = players.Count; i < 0; i--)
+            {
+                Player ply = players[i];
+                if (!Main.InstantKill.Remove(ply))
+                    players.Remove(ply);
+            }
+
+            response = $"All the followed player have been removed from InstantKill:\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/InstantKill/Remove.cs:48]:
```suggestion
            response = $"Every player has been removed from InstantKill:\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Inventory/Drop.cs [hunk 8]
```diff
@@ -33,27 +32,18 @@ protected override bool ExecuteParent(ArraySegment<string> arguments, ICommandSe
...
-                    response = $"All items from {pl.Nickname}'s inventory has been dropped";
-                    return true;
+                response = $"Player not found: {arguments.At(0)}";
+                return false;
             }
+
+            foreach (Player p in players)
+                p.DropItems();
+
+            response = $"All item has been dropped for all the followed players: \n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Inventory/Drop.cs:45]:
```suggestion
            response = $"All items have been dropped from the following players: \n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Jail.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                    DoUnJail(ply);
+                else
+                {
+                    if (Main.JailedPlayers.ContainsKey(ply.UserId))
+                        DoUnJail(ply);
+                    else
+                        DoJail(ply);
+                }
+            }
+            response = $"I am lazy to make the response now\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Jail.cs:65]:
```suggestion
            response = $"Jail command has run successfully.\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Position.cs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+                        return false;
+                    }
+                    if (!float.TryParse(arguments.At(3), out float yval))
+                    {
+                        response = $"Invalid value for x position: {arguments.At(3)}";
+                        return false;
+                    }
+                    if (!float.TryParse(arguments.At(4), out float zval))
+                    {
+                        response = $"Invalid value for x position: {arguments.At(4)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Position.cs:69]:
```suggestion
                        response = $"Invalid value for x position: {arguments.At(2)}";
                        return false;
                    }
                    if (!float.TryParse(arguments.At(3), out float yval))
                    {
                        response = $"Invalid value for y position: {arguments.At(3)}";
                        return false;
                    }
                    if (!float.TryParse(arguments.At(4), out float zval))
                    {
                        response = $"Invalid value for z position: {arguments.At(4)}";
```

File: AdminTools/Commands/Position.cs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                case PositionModifier.Get:
+                    if (arguments.Count != 2)
+                    {
+                        response = "Usage: position (all / *) (get)";
+                        return false;
+                    }
+                    StringBuilder positionBuilder = StringBuilderPool.Shared.Rent();
+                    if (players.IsEmpty())
+                    {
+                        response = "There are no players currently found";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Position.cs:92]:
```suggestion
                        response = "No players match your query.";
```

File: AdminTools/Commands/Position.cs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                        response = $"Invalid value for position: {arguments.At(3)}";
+                        return false;
+                    }
+                    switch (axis)
+                    {
+                        case VectorAxis.X:
+                            foreach (Player ply in players)
+                                ply.Position = new Vector3(ply.Position.x + val, ply.Position.y, ply.Position.z);
+
+                            response = $"Every player's x position has been added by {val}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Position.cs:135]:
```suggestion
                            response = $"Every player's x position has been nudged by {val}";
```

File: AdminTools/Commands/Position.cs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+                            return true;
+                        case VectorAxis.Y:
+                            foreach (Player ply in players)
+                                ply.Position = new Vector3(ply.Position.x, ply.Position.y + val, ply.Position.z);
+
+                            response = $"Every player's y position has been added by {val}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Position.cs:141]:
```suggestion
                            response = $"Every player's y position has been nudge by {val}";
```

File: AdminTools/Commands/Position.cs [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+                            return true;
+                        case VectorAxis.Z:
+                            foreach (Player ply in players)
+                                ply.Position = new Vector3(ply.Position.x, ply.Position.y, ply.Position.z + val);
+
+                            response = $"Every player's z position has been added by {val}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Position.cs:147]:
```suggestion
                            response = $"Every player's z position has been nudged by {val}";
```

File: AdminTools/Commands/PryGate/PryGates.cs [hunk 15]
```diff
@@ -81,63 +82,43 @@ protected override bool ExecuteParent(ArraySegment<string> arguments, ICommandSe
...
-
-                    if (Main.PryGateHubs.Contains(plyr))
+                    response = string.Empty;
+                    foreach (Player ply in players)
                     {
-                        Main.PryGateHubs.Remove(plyr);
-                        response = $"Player \"{plyr.Nickname}\" cannot pry gates open now";
+                        if (Main.PryGate.Remove(ply))
+                        {
+                            response += $"Player \"{ply.Nickname}\" cannot pry gates open now";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/PryGate/PryGates.cs:97]:
```suggestion
                            response += $"Player \"{ply.Nickname}\" can no longer pry gates open";
```

File: AdminTools/Commands/PryGate/PryGates.cs [hunk 16]
```diff
@@ -81,63 +82,43 @@ protected override bool ExecuteParent(ArraySegment<string> arguments, ICommandSe
...
+                            continue;
+                        }
+                        response += $"Player {ply.Nickname} does not have the ability to pry gates open";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/PryGate/PryGates.cs:100]:
```suggestion
                        response += $"Player {ply.Nickname} doesn't have the ability to pry gates open";
```

File: AdminTools/Commands/PryGate/PryGates.cs [hunk 17]
```diff
@@ -81,63 +82,43 @@ protected override bool ExecuteParent(ArraySegment<string> arguments, ICommandSe
...
-                        response = $"Player \"{pl.Nickname}\" can now pry gates open";
-                        return true;
-                    }
-                    else
-                    {
-                        Main.PryGateHubs.Remove(pl);
-                        response = $"Player \"{pl.Nickname}\" cannot pry gates open now";
-                        return true;
-                    }
+                    response = "The ability to pry gates open is on for all players now";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/PryGate/PryGates.cs:120]:
```suggestion
                    response = "Every player can now pry gates open.";
```

File: AdminTools/Commands/RandomTeleport.cs [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+                response = $"Player not found: {arguments.At(0)}";
+                return false;
+            }
+
+            foreach (Player ply in players)
+            {
+                ply.RandomTeleport(typeof(Room));
+            }
+
+            response = $"Followed player was teleported to a random room in the facility:\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/RandomTeleport.cs:47]:
```suggestion
            response = $"All specified players have been teleported to a random room in the facility:\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Regeneration.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                response = $"Player not found: {arguments.At(0)}";
+                return false;
+            }
+
+            AnimationCurve animationCurve = AnimationCurve.Constant(0f, duration, rate);
+            RegenerationProcess reg = new(animationCurve, 1f, 1f);
+            foreach (Player player in players)
+                UsableItemsController.GetHandler(player.ReferenceHub).ActiveRegenerations.Add(reg);
+            response = $"All the followed player has been regenerated:\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Regeneration.cs:59]:
```suggestion
            response = $"The specified player has been regenerated:\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Rocket.cs [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+{
+    [CommandHandler(typeof(RemoteAdminCommandHandler))]
+    [CommandHandler(typeof(GameConsoleCommandHandler))]
+    public class Rocket : ICommand, IUsageProvider
+    {
+        public string Command { get; } = "rocket";
+
+        public string[] Aliases { get; } = Array.Empty<string>();
+
+        public string Description { get; } = "Sends players high in the sky and explodes them";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Rocket.cs:19]:
```suggestion
        public string Description { get; } = "Turns the player into a firework.";
```

File: AdminTools/Commands/Rocket.cs [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+            if (!float.TryParse(arguments.At(1), out float speed) && speed <= 0)
+            {
+                response = $"Speed argument should be a positive number : {arguments.At(1)}";
+                return false;
+            }
+
+            foreach (Player ply in players)
+                Timing.RunCoroutine(DoRocket(ply, speed));
+
+            response = $"All the followed player has been rocketed into the sky\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Rocket.cs:53]:
```suggestion
            response = $"The specified players have been turned into fireworks\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Scale.cs [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+{
+    [CommandHandler(typeof(RemoteAdminCommandHandler))]
+    [CommandHandler(typeof(GameConsoleCommandHandler))]
+    public class Scale : ICommand, IUsageProvider
+    {
+        public string Command { get; } = "scale";
+
+        public string[] Aliases { get; } = Array.Empty<string>();
+
+        public string Description { get; } = "Scales all users or a user by a specified value";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Scale.cs:18]:
```suggestion
        public string Description { get; } = "Scales a specific player or all players.";
```

File: AdminTools/Commands/Scale.cs [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+            {
+                response = $"Invalid value for scale: {arguments.At(1)}";
+                return false;
+            }
+
+            Vector3 size = Vector3.one * scale;
+            foreach (Player ply in players)
+                ply.Scale = size;
+
+            response = $"Scale has been set to {size} for the followed player:\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Scale.cs:54]:
```suggestion
            response = $"The specified player's size has been set to {size}:\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/Size.cs [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+                response = $"Invalid value for z size: {arguments.At(3)}";
+                return false;
+            }
+            Vector3 size = new(x, y, z);
+            foreach (Player ply in players)
+            {
+                ply.Scale = size;
+            }
+
+            response = $"Size has been set to {size} for the followed player:\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/Size.cs:73]:
```suggestion
            response = $"The specified player's size has been set to {size}:\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Commands/TargetGhost.cs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+                {
+                    if (sourcePlayer.Role.Is(out FpcRole role))
+                    {
+                        if (!role.IsInvisibleFor.Add(victim))
+                            role.IsInvisibleFor.Remove(victim);
+                    }
+                }
+            }
+
+            response = $"Done modifying who can see who";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/TargetGhost.cs:65]:
```suggestion
            response = $"Finshed Ghostbusting.";
```

Comment (ALEXWARELLC) [reply AdminTools/Commands/TargetGhost.cs:65]: I'll leave this one up to whether or not people are ok with it.

This was literally for fun bc this vc is quiet af.

File: AdminTools/Commands/TeleportX.cs [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+                response = $"Player not found: {arguments.At(0)}";
+                return false;
+            }
+
+            foreach (Player plyr in players)
+            {
+                plyr.Position = ply.Position;
+            }
+
+            response = $"All the followed player has been teleported to {ply.Nickname}({ply.Id}):\n{Extensions.LogPlayers(players)}";
```
Comment (ALEXWARELLC) [inline AdminTools/Commands/TeleportX.cs:55]:
```suggestion
            response = $"The specified players has been teleported to {ply.Nickname}({ply.Id}):\n{Extensions.LogPlayers(players)}";
```

File: AdminTools/Extensions.cs [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+
+            if (player.BadgeHidden && !tagsRead.Contains(userId))
+            {
+                tagsRead.Add(userId);
+                Log.Debug($"{player.Nickname}({player.UserId}) has added their tag hidden.");
+            }
+            else if (!player.BadgeHidden && tagsRead.Remove(userId))
+                Log.Debug($"{player.Nickname}({player.UserId}) has remove their tag hidden.");
+        }
+        public static void SpawnWorkbench(Player ply, Vector3 position, Vector3 rotation, Vector3 size, out int benchIndex)
```
Comment (ALEXWARELLC) [inline AdminTools/Extensions.cs:51]: Someone else can fix this bad english.

**Review by ALEXWARELLC — CHANGES_REQUESTED** [review]
Completed Review.

This was and is a nightmare.

**Pull request merged**